File: 22_专题/区间问题/区间操作/1272. 删除区间.ts
// 左闭右开区间
function removeInterval(intervals: number[][], toBeRemoved: number[]): number[][] {
  const [badStart, badEnd] = toBeRemoved
  const res: [number, number][] = []

  intervals.forEach(([start, end]) => {
    // 不重叠
    if (end < badStart || start > badEnd) {
      res.push([start, end])
    } else {
      if (start < badStart) res.push([start, badStart]) // 重叠
      if (badEnd < end) res.push([badEnd, end])
    }
  })

  return res
}

console.log(
  removeInterval(
    [
      [0, 2],
      [3, 4],
      [5, 7]
    ],
    [1, 6]
  )
)
// 输出：[[0,1],[6,7]]

console.log(
  removeInterval(
    [
      [-5, -4],
      [-3, -2],
      [1, 2],
      [3, 5],
      [8, 9]
    ],
    [-1, 4]
  )
)

File: 22_专题/区间问题/区间操作/1272. 删除区间.py
# 1272. 删除区间-左闭右开区间
# https://leetcode.cn/problems/remove-interval/

from typing import List


class Solution:
    def removeInterval(self, intervals: List[List[int]], toBeRemoved: List[int]) -> List[List[int]]:
        badStart, badEnd = toBeRemoved
        res = []
        for start, end in intervals:
            if end < badStart or start > badEnd:
                res.append([start, end])
            else:
                if start < badStart:
                    res.append([start, badStart])
                if end > badEnd:
                    res.append([badEnd, end])
        return res
